core/src/main/java/org/elasticsearch/action/support/indices/TransportNodeBroadcastAction.java

```diff
@@ -0,0 +1,375 @@
+     * @param responses        the per-node aggregated shard-level responses
+     * @param shardFailures    the exceptions corresponding to shard operationa failures
+     * @return the response
+     */
+    protected abstract Response newResponse(Request request, int totalShards, int successfulShards, int failedShards, List<NodeBroadcastResponse> responses, List<DefaultShardOperationFailedException> shardFailures);
+
+    /**
+     * Creates a new node-level request.
+     *
+     * @param nodeId  the node the request will be sent to
+     * @param request the underlying request
+     * @param shards  the shards the operation will be executed against
+     * @return the node-level request
+     */
+    protected abstract NodeBroadcastRequest newNodeRequest(String nodeId, Request request, List<ShardRouting> shards);
+
+    /**
+     * Creates an empty node-level response object (used for deserialization).
+     *
+     * @return an empty node-level response object
+     */
+    protected abstract NodeBroadcastResponse newNodeResponse();
+
+    /**
+     * Creates a new node-level response.
+     *
+     * @param nodeId           the node the shard-level operations were executed on
+     * @param totalShards      the total number of shards for which execution of the operation was attempted
+     * @param successfulShards the total number of shards for which the execution of the operation was successful
+     * @param results          the per-shard operation results
+     * @param exceptions       the exceptions corresponding to shard operation failures
+     * @return the per-node aggregated shard-level response
+     */
+    protected abstract NodeBroadcastResponse newNodeResponse(String nodeId, int totalShards, int successfulShards, List<ShardOperationResult> results, List<BroadcastShardOperationFailedException> exceptions);
+
+    /**
+     * Executes the shard-level operation. This method is called asynchronously once per shard.
+     *
+     * @param request      the node-level request
+     * @param shardRouting the shard on which to execute the operation
+     * @return the result of the shard-level operation for the shard
+     */
+    protected abstract ShardOperationResult shardOperation(NodeBroadcastRequest request, ShardRouting shardRouting);
+
+    /**
+     * Determines the shards on which this operation will be executed on. The operation is executed once per shard
+     * iterator and unlike TransportBroadcastAction does not fall back to the next shard in the iterator on failure.
+     *
+     * @param clusterState    the cluster state
+     * @param request         the underlying request
+     * @param concreteIndices the concrete indices on which to execute the operation
+     * @return the shards on which to execute the operation
+     */
+    protected abstract GroupShardsIterator shards(ClusterState clusterState, Request request, String[] concreteIndices);
```

Comment:
I think we should move away from GroupShardIterator and just use plain vanilla ShardIterator. This is just confusing and doesn't buy much.